File: src/server/app/mcp_oauth.py
"""User-level MCP OAuth API — connect, callback, disconnect, schema refresh.

Endpoints:
- POST   /api/v1/mcp/servers/{name}/oauth/start    → {authorize_url}
- GET    /api/v1/mcp/oauth/callback                → 302 back into the app
- DELETE /api/v1/mcp/servers/{name}/oauth          → disconnect
- POST   /api/v1/mcp/servers/{name}/oauth/refresh-schemas → host-side re-discovery

The callback carries NO session auth by design: it is the AS redirecting the
user's browser, on any worker — identity comes exclusively from the
single-use ``state`` record minted in phase 1.
"""

from __future__ import annotations

import logging

from fastapi import APIRouter, Body, HTTPException, Request, Response
from fastapi.responses import RedirectResponse

from src.config.env import SERVER_BASE_URL
from src.server.services.mcp_oauth import (
    McpOAuthError,
    McpServerMoved,
    McpServerNotFound,
    TokenUnavailable,
    complete_callback,
    disconnect_server,
    start_connect,
)
from src.server.services.mcp_oauth.connect import STATE_TTL_SECONDS
from src.server.services.mcp_oauth.redirects import DEFAULT_RETURN_TO, CallbackError
from src.server.utils.api import CurrentUserId, handle_api_exceptions

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/mcp", tags=["MCP OAuth"])

# CSRF binding cookie for the OAuth connect flow. Path-scoped to the callback,
# HttpOnly, SameSite=Lax (must survive the AS's top-level cross-site redirect —
# Strict would not send it). Secure only on an HTTPS deployment so http dev
# still works.
#
# The name is per-state (``mcp_oauth_cb_<state>``): a fixed name would let two
# concurrent connects from the same browser overwrite each other's nonce cookie
# (same name, same path), so the first flow's callback would read the second
# flow's nonce and fail state_mismatch. Keying on the single-use state isolates
# concurrent flows and lets each callback delete exactly its own cookie.
_OAUTH_COOKIE_PREFIX = "mcp_oauth_cb_"
_OAUTH_COOKIE_PATH = "/api/v1/mcp/oauth"
_OAUTH_COOKIE_SECURE = SERVER_BASE_URL.lower().startswith("https")


def _oauth_cookie_name(state: str) -> str:
    return f"{_OAUTH_COOKIE_PREFIX}{state}"


@router.post("/servers/{name}/oauth/start")
@handle_api_exceptions("start MCP OAuth connect", logger)
async def oauth_start(
    name: str,
    user_id: CurrentUserId,
    request: Request,
    response: Response,
    body: dict | None = Body(default=None),
) -> dict:
    return_to = (body or {}).get("return_to")
    # Coerced here rather than in the service: the body is an unvalidated dict,
    # and anything that is not a list of keys is not a consent decision. A
    # non-list becomes "no selection named", which the service refuses outright
    # for a brokerage rather than reading as either extreme -- nobody was asked,
    # so it is neither declining everything nor granting it.
    raw_capabilities = (body or {}).get("granted_capabilities")
    granted_capabilities = (
        [str(key) for key in raw_capabilities]
        if isinstance(raw_capabilities, list)
        else None
    )
    try:
        started = await start_connect(
            user_id,
            name,
            return_to=return_to,
            # The browser's Origin is where the UI lives; the callback later
            # redirects there, since its own origin is the API on split ports.
            web_origin=request.headers.get("origin"),
            # A desktop shell offering its own listener, for an AS that refuses
            # a hosted callback. Bounded to loopback by the service and ignored
            # when it is anything else, so an unrecognised value degrades to the
            # ordinary flow rather than failing the request.
            loopback_redirect=(body or {}).get("redirect_uri"),
            # The address the page drew this row from. A connect can carry a
            # question that was asked of a particular server, and the row is the
            # user's to edit from another tab in between; naming it here is what
            # lets that question be a gate. Absent from an older page, which then
            # behaves as it always did.
            expected_url=(body or {}).get("expected_url"),
            # The capability groups agreed to in the dialog that preceded this
            # call. Intersected with the vendor's real groups by the service, so
            # a page that names none grants none.
            granted_capabilities=granted_capabilities,
        )
    except McpServerNotFound as e:
        raise HTTPException(status_code=404, detail=str(e))
    except McpServerMoved as e:
        # 409 rather than 422: the request is well formed and was right when the
        # page was drawn. What it conflicts with is the row's current state, and
        # the client tells the two apart to say "reload" rather than "invalid".
        raise HTTPException(status_code=409, detail=str(e))
    except McpOAuthError as e:
        raise HTTPException(status_code=422, detail=str(e))
    # Bind the callback to THIS browser: the nonce goes only into an HttpOnly
    # cookie, never the JSON body. The callback requires it back, so a stolen
    # (state, code) replayed in another browser has no matching cookie. Empty
    # on a loopback callback, where the cookie provably cannot come back — see
    # redirects.callback_is_loopback.
    if started.browser_nonce:
        response.set_cookie(
            _oauth_cookie_name(started.state),
            started.browser_nonce,
            # The state record it binds to is what expires the flow; a cookie
            # outliving it would only ever be sent at a state that is gone.
            max_age=STATE_TTL_SECONDS,
            path=_OAUTH_COOKIE_PATH,
            httponly=True,
            secure=_OAUTH_COOKIE_SECURE,
            samesite="lax",
        )
    return {
        "authorize_url": started.authorize_url,
        # Both of these exist for a desktop shell that armed a loopback listener
        # before this request, and neither is a secret: `state` is already in the
        # authorize URL the caller is about to be sent to, and `redirect_uri` is
        # the value it just offered. The shell needs `state` to tell its own
        # callback from anything else that reaches the port, and `redirect_uri`
        # to notice that a build which does not read the field left it on the
        # hosted callback -- the one failure that otherwise looks like success
        # right up until nothing arrives. The nonce stays cookie-only.
        "state": started.state,
        "redirect_uri": started.redirect_uri,
    }


@router.get("/oauth/callback")
async def oauth_callback(
    request: Request,
    state: str | None = None,
    code: str | None = None,
    iss: str | None = None,
    error: str | None = None,
    error_description: str | None = None,
) -> RedirectResponse:
    """AS redirect target. Always answers a redirect — never an error page."""
    # The nonce cookie is named for this flow's state; with no state there is no
    # cookie to read and complete_callback will reject on the missing state.
    cookie_name = _oauth_cookie_name(state) if state else None
    try:
        target = await complete_callback(
            state=state,
            code=code,
            iss=iss,
            error=error,
            error_description=error_description,
            browser_nonce=(
                request.cookies.get(cookie_name) if cookie_name else None
            ),
        )
    except Exception:
        logger.exception("[mcp_oauth] callback crashed")
        target = f"{DEFAULT_RETURN_TO}?mcp_error={CallbackError.INTERNAL}"
    # 303: the browser must GET the app route regardless of how it got here.
    resp = RedirectResponse(url=target, status_code=303)
    # The nonce is single-use — clear this flow's cookie so a later navigation
    # can't resend it.
    if cookie_name:
        resp.delete_cookie(cookie_name, path=_OAUTH_COOKIE_PATH)
    return resp


@router.delete("/servers/{name}/oauth")
@handle_api_exceptions("disconnect MCP OAuth", logger)
async def oauth_disconnect(name: str, user_id: CurrentUserId) -> dict:
    found = await disconnect_server(user_id, name)
    if not found:
        raise HTTPException(status_code=404, detail="No OAuth connection found")
    return {"ok": True}


@router.post("/servers/{name}/oauth/refresh-schemas")
@handle_api_exceptions("refresh MCP OAuth schemas", logger)
async def oauth_refresh_schemas(name: str, user_id: CurrentUserId) -> dict:
    from src.server.services.mcp_oauth.discovery import refresh_user_tool_schemas

    try:
        row = await refresh_user_tool_schemas(user_id, name)
    except TokenUnavailable as e:
        raise HTTPException(
            status_code=409, detail=f"Connection unusable: {e.reason}"
        )
    return {
        "server_name": row["server_name"],
        "status": row["status"],
        "error": row.get("error") or "",
        "tool_count": len(row.get("tools") or []),
        "discovered_at": row.get("discovered_at"),
    }

File: web/src/pages/ChatAgent/utils/api/mcpOauth.ts
/**
 * The user-level OAuth connect flow driven from the Plugins page, including the
 * desktop shell's loopback callback. Per-server OAuth *state* lives in mcp.ts.
 */
import { api } from '@/api/client';
import {
  beginMcpOAuth,
  bindMcpOAuth,
  canBeginMcpOAuth,
  cancelMcpOAuth,
  isDesktopShell,
  type McpOAuthFlow,
} from '@/lib/desktop';

const OAUTH_CALLBACK_PATH = '/api/v1/mcp/oauth/callback';

/**
 * Offer the desktop shell's loopback listener as this flow's redirect target.
 *
 * Some authorization servers allowlist only the native-app profile and reject a
 * hosted callback outright, which is a wall a browser cannot get past. The shell
 * can, so it is asked first and its answer travels with the request: the backend
 * binds the redirect target into the flow at this moment and checks it again at
 * the token exchange, which is why this has to be settled here rather than
 * rewritten in the navigation later.
 *
 * Undefined means there is no listener on offer, and the flow proceeds the way
 * it does in a browser.
 */
async function loopbackFlow(): Promise<McpOAuthFlow | undefined> {
  // Built outside the ask, because the two failures are not the same fault and
  // must not answer alike: a base URL this cannot parse is a misconfigured
  // deployment, and reporting it as "no listener" sends whoever is debugging it
  // to the shell, which was fine. `beginMcpOAuth` swallows its own.
  const callback = new URL(
    OAUTH_CALLBACK_PATH,
    api.defaults.baseURL || window.location.origin,
  ).toString();
  return beginMcpOAuth(callback);
}

/** Phase 1's answer. `state` and `redirect_uri` are for the desktop path. */
export interface McpOauthStart {
  authorize_url: string;
  /** This flow's OAuth `state`; absent from a build older than the field. */
  state?: string;
  /** The callback the flow was really minted against; see `startMcpOauth`. */
  redirect_uri?: string;
}

/**
 * A connect that had to come home through the shell could not be set up that way.
 *
 * Carries the reason rather than a sentence because the four are not the same
 * fault and do not have the same remedy, and the words belong to whoever is
 * rendering them: this module has no translator. `shell-outdated` and
 * `no-listener` are the pair most worth keeping apart -- one is answered by
 * updating the app and the other by quitting a second copy of it, and a reader
 * given the wrong one of those goes looking for a window that is not open.
 */
export class LoopbackRequiredError extends Error {
  constructor(
    readonly reason: 'shell-outdated' | 'no-listener' | 'not-minted' | 'not-bound',
  ) {
    super(`loopback callback unavailable: ${reason}`);
    this.name = 'LoopbackRequiredError';
  }
}

/** What a connect needs to know beyond which row it is for. */
export interface StartMcpOauthOptions {
  /**
   * This vendor's authorization server will not accept the hosted callback at
   * all, so the flow can only come home through the desktop shell's listener.
   */
  vendorRefusesHostedCallback?: boolean;
  /**
   * The address the caller drew the row from, checked against the row before
   * anything starts.
   *
   * The row is the user's to edit from any tab, so what the page asked them
   * about and what the connect would do can be two different servers -- and a
   * vendor that allows one connected AI platform per account makes that
   * difference expensive. Naming the address is what turns the question the
   * page asked into a gate. A caller that names none has nothing to be wrong
   * about and is let through as before.
   */
  expectedUrl?: string | null;
  /**
   * Whether the caller still wants this connect, asked once there is a URL to
   * navigate to. Answering false gives back whatever this armed and returns
   * null.
   *
   * The question belongs here rather than to the caller because the thing that
   * has to be released is not the caller's to see: arming happens inside this
   * function, before either round trip below, and the flow id never leaves it.
   * A caller that took the connect back in the meantime would otherwise be
   * holding a listener it has no handle on, and the shell would run the flow's
   * full timeout and then raise the window over a connect the user stopped ten
   * minutes earlier.
   */
  stillWanted?: () => boolean;
  /**
   * The capability groups the user agreed this connection may carry, or
   * undefined for a connect that was never asked.
   *
   * The empty array and undefined are different answers and both travel as they
   * are: `[]` is a brokerage granted nothing, undefined is a server we curate no
   * groups for and hold no policy over. Collapsing them would make "declined
   * everything" and "not a brokerage" the same request, and only one of those
   * may reach a vendor's whole tool list.
   */
  grantedCapabilities?: string[];
}

/**
 * Phase 1 of the connect flow; the caller navigates to `authorize_url`.
 *
 * Everything past `returnTo` is named rather than positional. Three of the four
 * are optional and two of them are a boolean beside a callback, which is the
 * shape an argument gets silently passed in the wrong slot.
 */
export async function startMcpOauth(
  name: string,
  returnTo = '/plugins',
  {
    vendorRefusesHostedCallback = false,
    expectedUrl,
    stillWanted,
    grantedCapabilities,
  }: StartMcpOauthOptions = {},
): Promise<McpOauthStart | null> {
  // Asked unconditionally, because asking is free: `loopbackFlow` answers
  // undefined wherever there is no shell, so this arms a listener exactly where
  // one exists and changes nothing in a browser.
  const flow = await loopbackFlow();
  // Whether its absence is fatal is the real question, and it is fatal in two
  // separate situations.
  //
  // A vendor that refuses the hosted callback has no other way home anywhere.
  //
  // And inside the shell, NOTHING has another way home -- whatever the vendor
  // accepts, and whatever this shell is old enough to offer. The authorize URL
  // is an external navigation, so it leaves for the system browser, and the
  // callback then lands in a browser that never received this flow's nonce
  // cookie -- the cookie was set on the request this window made. The backend
  // refuses it as a state mismatch, in a browser the user may not even be
  // signed into, while this window sits on "connecting" because its navigation
  // was cancelled and it never unloaded. Degrading to the hosted callback there
  // is a guaranteed failure nobody is told about, which is worse than not
  // starting.
  if (!flow) {
    // Being IN the shell is the question, not what this shell can do. A build
    // that predates the loopback channel has no method to find, and reading
    // that as "not the desktop app" is how it ends up on the browser path --
    // the one path that cannot work here.
    if (isDesktopShell() && !canBeginMcpOAuth()) {
      throw new LoopbackRequiredError('shell-outdated');
    }
    if (vendorRefusesHostedCallback || isDesktopShell()) {
      throw new LoopbackRequiredError('no-listener');
    }
  }
  try {
    const { data } = await api.post<McpOauthStart>(
      `/api/v1/mcp/servers/${name}/oauth/start`,
      // One guard, and it is `loopbackFlow` above: every way of not having a
      // URI already arrives here as undefined. A second truthiness check would
      // read as belt and braces and is really a place for the two to disagree.
      {
        return_to: returnTo,
        ...(flow ? { redirect_uri: flow.redirectUri } : {}),
        ...(expectedUrl ? { expected_url: expectedUrl } : {}),
        // Presence, not truthiness, unlike the two above: an empty selection is
        // a decision the user made and has to be sent, and it is the one value
        // here that a shorthand check would drop.
        ...(grantedCapabilities !== undefined
          ? { granted_capabilities: grantedCapabilities }
          : {}),
      },
    );
    if (flow) {
      // Did the flow actually get minted against the listener we armed? A build
      // that predates the field ignores it and mints the hosted callback, and
      // so does this one when the value fails its loopback check -- both answer
      // 200, so without asking, the only symptom is a listener holding for five
      // minutes on a code that was never coming, and then raising the window.
      // For the vendors this path exists for there is no error even then: their
      // consent screen simply never redirects anywhere we can hear.
      if (data.redirect_uri !== flow.redirectUri || !data.state) {
        throw new LoopbackRequiredError('not-minted');
      }
      // Only now can the shell recognise this flow's callback. Before it, an
      // armed listener accepts nothing at all -- and a second connect started
      // in this window while the first was still minting takes the slot, so a
      // refusal here is this flow losing that race, not a broken shell.
      if (!(await bindMcpOAuth(flow.flowId, data.state))) {
        throw new LoopbackRequiredError('not-bound');
      }
    }
    // Asked last, with the listener armed and bound and the URL in hand, since
    // that is the widest the window gets: everything above is a round trip the
    // user could have pressed Cancel through. Not raised as a failure -- there
    // is nothing wrong here and nothing to tell them about a connect they are
    // the one who stopped.
    if (stillWanted && !stillWanted()) {
      if (flow) await cancelMcpOAuth(flow.flowId);
      return null;
    }
    return data;
  } catch (err) {
    // Asking armed a listener, and this is the answer that it was for nothing.
    // The order is forced -- the redirect_uri has to be in hand before the
    // request that binds it -- so the only way to keep arming honest is to say
    // when the flow it was armed for did not happen. By id: this window may
    // have started another connect since, and that one is still live.
    if (flow) await cancelMcpOAuth(flow.flowId);
    throw err;
  }
}

export async function disconnectMcpOauth(name: string) {
  const { data } = await api.delete(`/api/v1/mcp/servers/${name}/oauth`);
  return data as { ok: boolean };
}

export interface McpOauthSchemaRefreshResult {
  server_name: string;
  status: string;
  error: string;
  tool_count: number;
  discovered_at: string | null;
}

/** Host-side re-discovery of an OAuth server's tool schemas. */
export async function refreshMcpOauthSchemas(
  name: string,
): Promise<McpOauthSchemaRefreshResult> {
  const { data } = await api.post<McpOauthSchemaRefreshResult>(
    `/api/v1/mcp/servers/${name}/oauth/refresh-schemas`,
  );
  return data;
}
